src/raftstore/coprocessor/region_info_accessor.rs

```diff
@@ -0,0 +1,724 @@
+                return;
+            }
+            // Insert new entry to `region_ranges`.
+            self.region_ranges.insert(end_key, region.get_id());
+        }
+
+        if need_create_new_region {
+            // It's a new region. Create it.
+            self.create_region(region);
+        }
+    }
+
+    fn handle_destroy_region(&mut self, region: Region) {
+        if let Some(mut removed_region_info) = self.regions.remove(&region.get_id()) {
+            let removed_region = removed_region_info.region;
+            assert_eq!(removed_region.get_id(), region.get_id());
+
+            // Ignore if it's a stale message
+            let removed_version = removed_region.get_region_epoch().get_version();
+            let removed_conf = removed_region.get_region_epoch().get_conf_ver();
+            if removed_version > region.get_region_epoch().get_version()
+                || removed_conf > region.get_region_epoch().get_conf_ver()
+            {
+                // It shouldn't be deleted. Put it back.
+                removed_region_info.region = removed_region;
+                self.regions.insert(region.get_id(), removed_region_info);
+                return;
+            }
+
+            assert_eq!(removed_region.get_id(), region.get_id());
+            let end_key = data_end_key(removed_region.get_end_key());
+
+            let removed_id = self.region_ranges.remove(&end_key).unwrap();
+            assert_eq!(removed_id, region.get_id());
+        } else {
+            // It's possible that the region is already removed because it's end_key is used by
+            // another newer region.
+            debug!(
+                "region_collection: destroying region {} but it doesn't exist",
+                region.get_id()
+            )
+        }
+    }
+
+    fn handle_role_change(&mut self, region: Region, new_role: StateRole) {
+        let region_id = region.get_id();
+        if self.regions.get(&region_id).is_none() {
+            warn!("region_collection: role change on region {} but the region doesn't exist. create it.", region_id);
+            self.create_region(region);
+        }
+
+        if let Some(r) = self.regions.get_mut(&region_id) {
+            r.role = new_role;
+        }
+    }
+
+    fn handle_raftstore_event(&mut self, event: RaftStoreEvent) {
+        match event {
+            RaftStoreEvent::CreateRegion { region } => {
+                self.handle_create_or_update_region(region);
+            }
+            RaftStoreEvent::UpdateRegion { region } => {
+                self.handle_create_or_update_region(region);
+            }
+            RaftStoreEvent::DestroyRegion { region } => {
+                self.handle_destroy_region(region);
+            }
+            RaftStoreEvent::RoleChange { region, role } => {
+                self.handle_role_change(region, role);
+            }
+        }
+    }
+}
+
+impl Runnable<RegionCollectionMsg> for RegionCollection {
```

Comment:
Impl runnable for collection is weird.

Reply:
So we should give `RegionCollection` another name?

Reply:
How about `RegionCollector`?